File: day2-functions-extensions-lambdas/java/FunctionExample.java
/**
 * Day 2 - Functions in Java vs Kotlin
 *
 * Java:
 *   - Method overloading (no named params, no default args)
 *   - varargs for variable arguments
 *   - Single-expression methods (Java 14+ with return)
 *
 * Kotlin:
 *   - Named arguments
 *   - Default parameter values
 *   - Single-expression functions
 *   - Infix functions
 *   - Extension functions
 *   - Higher-order functions
 */
import java.util.List;
import java.util.ArrayList;

public class FunctionExample {

    // ═══════════════════════════════════════════════════════════════
    // 1. JAVA FUNCTIONS (methods)
    // ═══════════════════════════════════════════════════════════════

    // Java: Must specify return type, parameter types
    public static String greet(String name) {
        return "Hello, " + name + "!";
    }

    // Java: Method overloading (different parameter list)
    public static String greet(String firstName, String lastName) {
        return "Hello, " + firstName + " " + lastName + "!";
    }

    // Java: Overloading with different types
    public static String greet(int times, String name) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < times; i++) {
            sb.append("Hello, ").append(name).append("! ");
        }
        return sb.toString().trim();
    }

    // ═══════════════════════════════════════════════════════════════
    // 2. JAVA: NO NAMED PARAMETERS!
    // ═══════════════════════════════════════════════════════════════

    // Java: Arguments passed by position ONLY
    // createPerson("Alice", 25, "alice@email.com")
    // If you want to skip a param, you MUST overload or use builder pattern

    // Java workaround: Builder pattern
    public static class PersonBuilder {
        private String name = "Unknown";
        private int age = 0;
        private String email = "";

        public PersonBuilder name(String name) { this.name = name; return this; }
        public PersonBuilder age(int age) { this.age = age; return this; }
        public PersonBuilder email(String email) { this.email = email; return this; }
        public String build() { return name + " (" + age + ") - " + email; }
    }

    // ═══════════════════════════════════════════════════════════════
    // 3. JAVA: NO DEFAULT PARAMETERS!
    // ═══════════════════════════════════════════════════════════════

    // Java: Must overload for default values
    public static String formatString(String text) {
        return formatString(text, "UPPER");  // Call overload with default
    }

    public static String formatString(String text, String mode) {
        return switch (mode) {
            case "UPPER" -> text.toUpperCase();
            case "LOWER" -> text.toLowerCase();
            default -> text;
        };
    }

    // Java: Many overloads needed!
    // print("hello")
    // print("hello", true)
    // print("hello", true, 2)
    // print("hello", true, 2, '\n')
    // Each combination needs a separate method!

    // ═══════════════════════════════════════════════════════════════
    // 4. JAVA: VARARGS (variable arguments)
    // ═══════════════════════════════════════════════════════════════

    public static int sum(int... numbers) {
        int total = 0;
        for (int n : numbers) {
            total += n;
        }
        return total;
    }

    // ═══════════════════════════════════════════════════════════════
    // 5. JAVA: SINGLE-EXPRESSION (Java 14+)
    // ═══════════════════════════════════════════════════════════════

    // Java 14+: yield for switch expressions
    // But no single-expression function syntax like Kotlin

    // ═══════════════════════════════════════════════════════════════
    // 6. JAVA: STATIC METHODS vs INSTANCE METHODS
    // ═══════════════════════════════════════════════════════════════

    // Java: Must use class name for static methods
    // FunctionExample.greet("Alice")

    // Java 19+: static import for cleaner calls
    // import static FunctionExample.greet;
    // greet("Alice");

    // ═══════════════════════════════════════════════════════════════
    // 7. COMPARISON TABLE
    // ═══════════════════════════════════════════════════════════════
    //
    //   Feature              Java 17                         Kotlin
    //   ──────────────────── ─────────────────────────────── ────────────────────────────
    //   Function syntax      public static String f(int x)  fun f(x: Int): String
    //   Named arguments      ❌ Not supported                ✅ f(x = 5)
    //   Default args         ❌ Must overload                ✅ fun f(x: Int = 5)
    //   Single expression    ❌ Must use body                ✅ fun f() = expr
    //   Varargs              int... args                     vararg args: Int
    //   Extension functions  ❌ Not possible                 ✅ fun String.f()
    //   Infix functions      ❌ Not possible                 ✅ infix fun f()
    //   Higher-order         ❌ Verbose (interfaces)         ✅ (Int) -> String
    //
    // ═══════════════════════════════════════════════════════════════

    public static void main(String[] args) {

        // Basic function calls
        System.out.println(greet("Alice"));
        System.out.println(greet("Alice", "Smith"));

        // No named parameters — must pass in order!
        // createPerson("Alice", 25, "alice@email.com")

        // Builder pattern workaround
        String person = new PersonBuilder()
            .name("Alice")
            .age(25)
            .email("alice@email.com")
            .build();
        System.out.println(person);

        // Default args workaround: overloading
        System.out.println(formatString("hello"));           // Uses overload
        System.out.println(formatString("hello", "lower"));  // Full params

        // Varargs
        System.out.println("Sum: " + sum(1, 2, 3, 4, 5));

        // ═══════════════════════════════════════════════════════════════
        // 8. KEY TAKEAWAYS
        // ═══════════════════════════════════════════════════════════════
        //
        // JAVA PROBLEMS:
        //   1. No named arguments → hard to read `f(true, false, true, false)`
        //   2. No default arguments → must overload for each combination
        //   3. No extension functions → must use utility classes
        //   4. No single-expression functions → verbose getters
        //   5. No infix functions → `map.get(key)` instead of `map[key]`
        //
        // KOTLIN SOLUTIONS:
        //   1. Named arguments: `f(a = true, c = true)` — skip b, d
        //   2. Default arguments: `fun f(a: Int = 1, b: Int = 2)`
        //   3. Extension functions: `fun String.isEmail()`
        //   4. Single-expression: `fun double(x: Int) = x * 2`
        //   5. Infix functions: `key map value`
        //
        // ═══════════════════════════════════════════════════════════════
    }
}

File: day2-functions-extensions-lambdas/kotlin/FunctionExample.kt
/**
 * Day 2 - Functions in Kotlin
 *
 * Kotlin functions have MANY advantages over Java:
 *   - Named arguments
 *   - Default parameter values
 *   - Single-expression functions
 *   - Extension functions
 *   - Higher-order functions
 *   - Infix functions
 */

// ═══════════════════════════════════════════════════════════════
// 1. BASIC FUNCTION SYNTAX
// ═══════════════════════════════════════════════════════════════

// Kotlin: fun keyword, type after parameter, return type after parentheses
fun greet(name: String): String {
    return "Hello, $name!"
}

// Java equivalent:
// public static String greet(String name) {
//     return "Hello, " + name + "!";
// }

// ═══════════════════════════════════════════════════════════════
// 2. SINGLE-EXPRESSION FUNCTIONS
// ═══════════════════════════════════════════════════════════════

// Kotlin: expression body (no return keyword needed!)
fun double(x: Int) = x * 2
fun isPositive(n: Int) = n > 0
fun add(a: Int, b: Int) = a + b

// Java equivalent:
// public static int double(int x) { return x * 2; }
// public static boolean isPositive(int n) { return n > 0; }

// ═══════════════════════════════════════════════════════════════
// 3. NAMED ARGUMENTS (Java doesn't have this!)
// ═══════════════════════════════════════════════════════════════

data class Person(val name: String, val age: Int, val email: String)

// Java: createPerson("Alice", 25, "alice@email.com")  — must pass in order
// Kotlin: can name arguments in any order!

// ═══════════════════════════════════════════════════════════════
// 4. DEFAULT PARAMETER VALUES (Java doesn't have this!)
// ═══════════════════════════════════════════════════════════════

fun formatString(text: String, mode: String = "UPPER"): String {
    return when (mode) {
        "UPPER" -> text.uppercase()
        "LOWER" -> text.lowercase()
        else -> text
    }
}

// Java equivalent:
// public static String formatString(String text) {
//     return formatString(text, "UPPER");  // Overload needed!
// }
// public static String formatString(String text, String mode) { ... }

// More complex example with multiple defaults
fun createUrl(
    host: String = "localhost",
    port: Int = 8080,
    path: String = "/",
    secure: Boolean = false
): String {
    val protocol = if (secure) "https" else "http"
    return "$protocol://$host:$port$path"
}

// ═══════════════════════════════════════════════════════════════
// 5. VARARG (variable arguments)
// ═══════════════════════════════════════════════════════════════

fun sum(vararg numbers: Int): Int {
    return numbers.sum()
}

// Can also use spread operator *
fun printNumbers(vararg numbers: Int) {
    println(numbers.joinToString(", "))
}

// ═══════════════════════════════════════════════════════════════
// 6. SINGLE-RETURN vs MULTIPLE RETURNS
// ═══════════════════════════════════════════════════════════════

// Java: Can only return one value
// Kotlin: Can return Pair or use data class for multiple returns

fun divide(a: Int, b: Int): Pair<Int, Int> {
    return Pair(a / b, a % b)  // quotient and remainder
}

// Better: use data class for named returns
data class DivideResult(val quotient: Int, val remainder: Int)

fun divideBetter(a: Int, b: Int): DivideResult {
    return DivideResult(a / b, a % b)
}

// ═══════════════════════════════════════════════════════════════
// 7. LOCAL FUNCTIONS (Java can't do this!)
// ═══════════════════════════════════════════════════════════════

fun processOrder(orderId: String) {
    // Local function — defined inside another function
    fun validate(id: String): Boolean {
        return id.isNotBlank() && id.startsWith("ORD-")
    }

    fun calculateTotal(items: List<Double>): Double {
        return items.sum()
    }

    if (validate(orderId)) {
        val items = listOf(99.99, 49.99, 29.99)
        println("Order $orderId total: $${calculateTotal(items)}")
    } else {
        println("Invalid order: $orderId")
    }
}

// Java equivalent: Extract to separate methods (can't define inside method)

// ═══════════════════════════════════════════════════════════════
// 8. FUNCTION OVERLOADING (same as Java)
// ═══════════════════════════════════════════════════════════════

fun display(text: String) = println("String: $text")
fun display(number: Int) = println("Int: $number")
fun display(list: List<*>) = println("List: ${list.size} items")

// But Kotlin prefers default args over overloading!

// ═══════════════════════════════════════════════════════════════
// MAIN FUNCTION
// ═══════════════════════════════════════════════════════════════

fun main() {

    // Basic function
    println(greet("Alice"))

    // Single-expression functions
    println("Double 5: ${double(5)}")
    println("Is positive? ${isPositive(3)}")

    // Named arguments — can skip defaults!
    println(formatString("hello"))                    // Uses default "UPPER"
    println(formatString("hello", mode = "LOWER"))   // Named arg, skip nothing
    println(formatString("hello", mode = "title"))   // Named arg

    // Named arguments in any order
    val url1 = createUrl()                                        // All defaults
    val url2 = createUrl(port = 9090, path = "/api")             // Custom port and path
    val url3 = createUrl(secure = true, host = "example.com")    // Custom host, secure
    println("URL 1: $url1")
    println("URL 2: $url2")
    println("URL 3: $url3")

    // Varargs
    println("Sum: ${sum(1, 2, 3, 4, 5)}")
    printNumbers(1, 2, 3)

    // Multiple returns
    val (quotient, remainder) = divide(17, 5)
    println("17 / 5 = $quotient remainder $remainder")

    val result = divideBetter(17, 5)
    println("17 / 5 = ${result.quotient} remainder ${result.remainder}")

    // Local functions
    processOrder("ORD-001")
    processOrder("INVALID")

    // Overloading
    display("Hello")
    display(42)
    display(listOf(1, 2, 3))

    // ═══════════════════════════════════════════════════════════════
    // COMPARISON SUMMARY
    // ═══════════════════════════════════════════════════════════════
    //
    // JAVA                              KOTLIN
    // ────────────────────────────────  ──────────────────────────────
    // public int add(int a, int b)      fun add(a: Int, b: Int): Int
    // public int add(int a, int b,      fun add(a: Int, b: Int,
    //               int c)                      c: Int = 0): Int
    // add(1, 2)                         add(1, 2)
    // add(1, 2, 3)                      add(1, 2, 3)
    // add(1, 2) + default overload     add(1, 2)  // c defaults to 0!
    // No named args                     add(a = 1, b = 2)
    // No local functions                fun local() { ... }
    // Return one value                  return Pair(a, b) or data class
    //
    // ═══════════════════════════════════════════════════════════════
}
